// @sudoku/stores/roam.js
import { writable } from "svelte/store";
import { userGrid } from "./grid"

function createRoam() {
    const roam = writable({
        undoStack: [],
        redoStack: [],
        checkpoint: null
    });

    return {
        subscribe: roam.subscribe,

        record() {
            roam.update($roam => {
                const currentState = userGrid.getState();
                return {
                    undoStack: [...$roam.undoStack, currentState],
                    redoStack: [],
                    checkpoint: $roam.checkpoint
                };
            })
        },

        undo() {
            roam.update($roam => {
                if ($roam.undoStack.length === 0) return $roam;

                const currentState = userGrid.getState();
                userGrid.setAll($roam.undoStack[$roam.undoStack.length - 1]);

                return {
                    undoStack: $roam.undoStack.slice(0, -1),
                    redoStack: [...$roam.redoStack, currentState],
                    checkpoint: $roam.checkpoint
                };
            })
        },

        redo() {
            roam.update($roam => {
                if ($roam.redoStack.length === 0) return $roam;

                const curentState = userGrid.getState();
                userGrid.setAll($roam.redoStack[$roam.redoStack.length - 1]);

                return {
                    undoStack: [...$roam.undoStack, curentState],
                    redoStack: $roam.redoStack.slice(0, -1),
                    checkpoint: $roam.checkpoint
                };
            })
        },

        save() {
            roam.update($roam => {
                const currentGrid = userGrid.getState();
                return {
                    ...$roam,
                    checkpoint: currentGrid
                };
            });
        },

        restore() {
            roam.update($roam => {
                if ($roam.checkpoint)
                    userGrid.setAll(JSON.parse(JSON.stringify(($roam.checkpoint))));
                return $roam;
            });
        },
    };
}

export const roam = createRoam();
