// @sudoku/stores/roam.js
import { writable } from "svelte/store";
import { userGrid } from "./grid"

function createRoam() {
    const roam = writable({
        undoStack: [],
        redoStack: []
    });

    return {
        subscribe: roam.subscribe,

        record() {
            roam.update($roam => {
                return {
                    undoStack: [...$roam.undoStack, userGrid.getState()],
                    redoStack: []
                };
            })
        },

        undo() {
            roam.update($roam => {
                if ($roam.undoStack.length === 0) return $roam;

                const currentState = userGrid.getState();
                userGrid.setAll($roam.undoStack[$roam.undoStack.length - 1]);

                return {
                    undoStack: $roam.undoStack.slice(0, -1),
                    redoStack: [...$roam.redoStack, currentState]
                };
            })
        },

        redo() {
            roam.update($roam => {
                if ($roam.redoStack.length === 0) return $roam;

                const curentState = userGrid.getState();
                userGrid.setAll($roam.redoStack[$roam.redoStack.length - 1]);

                return {
                    undoStack: [...$roam.undoStack, curentState],
                    redoStack: $roam.redoStack.slice(0, -1)
                };
            })
        }
    };
}

export const roam = createRoam();
