// @sudoku/strategy/strategy.js
import BoxScanner, {getBox, listToString} from '@sudoku/strategy/utils';

class StrategyCell {
    constructor(y, x, values) {

        //需要高亮的候选值
        this.highLightValues = [];

        for (let i = 0; i < values.length; i++) {
            this.highLightValues.push(values[i]);
        }

        this.y = y;
        this.x = x;
    }
}

class StrategyUnit {
    constructor(name) {
        this.strategyName = name;

        this.cells = [];  // StrategyCell的数组
    }

    addCell (y, x, values) {
        this.cells.push(new StrategyCell(y, x, values));
    }

    len() {
        return this.cells.length;
    }
}

class StrategyOperator
{
    constructor(name) {
        this.enabled = false;
        this.opName = name;
    }

    begin () {
        this.enabled = true;
    }

    excute (y, x, candidates) {
        if (!this.enabled) {
            console.log(this.opName() + " use excute() before begin()!");
        }
    }

    end () {
        this.enabled = false;
    }

    name() {
        return this.opName;
    }
}

class Strategy
{
    constructor(name, operator) {
        this.strategyName = name;
        this.operator = operator;
    }

    run (candidates, cellTrack, strategyUnits) {
        this.iterativeSearch(candidates);
        let ret = this.handle(candidates, cellTrack, strategyUnits);
        return ret;
    }

    //迭代查找
    iterativeSearch (candidates) {
        for (let y = 0; y < 9; y++) {
            this.operator.begin();
            for (let x = 0; x < 9; x++) {
                this.operator.excute(y, x, candidates);
            }
            load(this.operator.end()); 
        }

        for (let x = 0; x < 9; x++) {
            this.operator.begin();
            for (let y = 0; y < 9; y++) {
                this.operator.excute(y, x, candidates);
            }
            load(this.operator.end());
        }

        for (let i = 0; i < 9; i++) {
            let bs = new BoxScanner(0, 0, i);
            this.operator.begin();

            for (bs.begin(); !bs.end(); bs.next()) {
                this.operator.excute(bs.y, bs.x, candidates);
            }
            load(this.operator.end());
        }
    }

    //存储迭代查找的中间结果
    load (midData) {

    }

    //根据迭代查找结果进行匹配和候选值处理
    handle (candidates, cellTrack, strategyUnits) {
        return false;
    }

    name () {
        return this.strategyName;
    }

}

export {
    StrategyCell,
    StrategyUnit,
    StrategyOperator,
    Strategy,
};
