// @sudoku/strategy/utils.js
class BoxScanner
{
    constructor(y, x, id) {
        if (arguments.length == 2) {
            this.startY = y;
            this.startX = x;
    
            this.y = -1;
            this.x = -1;
        }
        else if (arguments.length == 3) {
            this.startY = Math.floor(id/3)*3;
            this.startX = (id%3)*3;
    
            this.y = -1;
            this.x = -1;
        }
    }

    begin () {
        this.y = this.startY;
        this.x = this.startX;
    }

    next () {
        if (this.x == this.startX + 2) {
            this.y++;
            this.x = this.startX;
        }
        else {
            this.x++;
        }
    }

    end () {
        if (this.x >= this.startX + 2 && this.y >= this.startY + 2) {
            return true;
        }
        else {
            return false;
        }
    }
}

//根据输入的坐标返回对应的是哪个box
export function getBox(y, x) {
    return parseInt(y/3) * 3 + parseInt(x/3);
}

export function listToString(list) {

    if (list.length === 0) {
        return '';
    }

    let ans = '';
    for (let i = 0; i < list.length - 1; i++) {
        ans = ans + list[i] + ' ';
    }
    ans = ans + list[list.length - 1];
    return ans
}

export default BoxScanner;
